Increase test coverage for fs/promises.js

humphd opened the pull request:
I was looking at [test coverage](https://coverage.nodejs.org/coverage-fdb35d89605de642/root/fs/promises.js.html) for the new things in `fs/promises.js` and wanted to add some more.  This adds test cases using `FileHandle` objects vs. only using paths or fds.

Due to https://github.com/nodejs/node/issues/19057, I'm unable to run coverage locally on macOS; apologies for not including updated coverage info.

Trott commented:
Hi @humphd! Welcome and thanks for the PR! It's possible that some/all of this is duplicated in https://github.com/nodejs/node/pull/19605. Might want to compare and remove anything that's duplicated because that one looks like it's going to land very soon. (As in I'm actually in the middle of landing it RIGHT NOW.)

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14064/

humphd commented:
@Trott OK, thanks, I'll compare with what's landed there and adjust this.

Trott commented:
(Lone CI failure is unrelated. Can be re-run if this PR isn't going to change, but if there are any changes, we'll have to re-run all of CI anyway. So, either way...)

humphd commented:
@Trott based on what was added in #19057, I've just reverted my changes to `test/parallel/test-fs-promises-writefile.js` and just included my additions to `test/parallel/test-fs-promises.js`, which are unmodified from my previous commit.  I've also rebased this.

Trott commented:
@nodejs/fs @nodejs/testing

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14103/

BridgeAR commented on test/parallel/test-fs-promises.js:
```diff
@@ -82,24 +82,44 @@ function verifyStatObject(stat) {
+        code: 'ERR_OUT_OF_RANGE',
+        type: RangeError
+      })(err);
+    }
```
As a general note: we can actually use the following instead:

```js
assert.rejects(
  // `mode` can't be > 0o777
  () => fchmod(handle, (0o777 + 1)),
  {
    code: 'ERR_OUT_OF_RANGE',
    name: 'RangeError [ERR_OUT_OF_RANGE]'
  }
);
```

BridgeAR commented on test/parallel/test-fs-promises.js:
```diff
@@ -147,6 +167,16 @@ function verifyStatObject(stat) {
+      await mkdtemp(1);
+    } catch (err) {
+      // mkdtemp() expects to get a string prefix.
+      console.log('err', err);
```
Nit: can you please remove the console.log here?

thefourtheye commented on test/parallel/test-fs-promises.js:
```diff
@@ -82,24 +82,44 @@ function verifyStatObject(stat) {
+    await handle.write(buf2);
+    const ret2 = await handle.read(Buffer.alloc(11), 0, 11, 0);
+    assert.strictEqual(ret2.bytesRead, 11);
+    assert.deepStrictEqual(ret2.buffer, buf);
```
Shouldn't this be compared with `buf2`?

thefourtheye replied:
Nit: Also, if we reduce magic numbers (like 11 in this case) it would be better.

thefourtheye commented on test/parallel/test-fs-promises.js:
```diff
@@ -82,24 +82,44 @@ function verifyStatObject(stat) {
     await fchmod(handle, 0o666);
+    handle.chmod(0o666);
+    try {
+      await fchmod(handle, (0o777 + 1));
```
It would be better, if did this for `chmod` as well.

BridgeAR commented:
@humphd would you be so kind and have a look at the comments? :-)

humphd commented:
@BridgeAR apologies for the delay.  I'll update this week.

humphd commented:
OK, I've rebased and done the following based on the reviews above:

* switched `try/catch`s I added to use `assert.rejects` instead
* removed stray `console.log`
* gotten rid of magic numbers
* corrected comparison of `buf` vs. `buf2`
* added test cases for `(0o777 + 1)` on `chmod` and `fchmod` as well.  I hope I've interpreted what was [requested above](https://github.com/nodejs/node/pull/19811#discussion_r180170141) correctly with this one.  Please let me know if I should alter this.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/14602/

TimothyGu commented on test/parallel/test-fs-promises.js:
```diff
@@ -82,24 +82,60 @@ function verifyStatObject(stat) {
+
     await chmod(dest, 0o666);
     await fchmod(handle, 0o666);
+    handle.chmod(0o666);
```
Should this be `await`ed?

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-commit/18242/

mhdawson commented:
OSX CI failure looks unrelated, opened https://github.com/nodejs/node/issues/20660

New CI run: https://ci.nodejs.org/job/node-test-pull-request/14795/

addaleax commented:
Landed in fcc46ee5a9ea9fdc66934cbba6a5e7b1486bb1aa
